scripts/common/rook-upgrade.sh

```diff
@@ -0,0 +1,523 @@
+}
+
+# rook_upgrade_addon_fetch will fetch all add-on versions from $from_version to $to_version.
+function rook_upgrade_addon_fetch() {
+    local from_version="$1"
+    local to_version="$2"
+
+    logStep "Downloading images required for Rook $from_version to $to_version upgrade"
+
+    if rook_upgrade_is_version_included "$from_version" "$to_version" "1.4" ; then
+        if ! rook_upgrade_addon_fetch_step "rookupgrade" "10to14" ; then
+            return 1
+        fi
+    fi
+
+    if [ "$(rook_upgrade_compare_rook_versions "$to_version" "1.4")" = "1" ]; then
+        local step=
+        while read -r step; do
+            if [ -z "$step" ] || [ "$step" = "0.0.0" ]; then
+                continue
+            fi
+            if ! rook_upgrade_addon_fetch_step "rook" "$step" ; then
+                return 1
+            fi
+        done <<< "$(rook_upgrade_step_versions "ROOK_STEP_VERSIONS[@]" "$(rook_upgrade_max_rook_version "1.4" "$from_version")" "$to_version")"
+    fi
+
+    logSuccess "Images loaded for Rook $from_version to $to_version upgrade"
+}
+
+# rook_upgrade_addon_fetch_step will fetch an individual add-on version.
+function rook_upgrade_addon_fetch_step() {
+    local addon="$1"
+    local version="$2"
+
+    if ! addon_exists "$addon" "$version" ; then
+        logFail "Rook version $version not found"
+        return 1
+    fi
+
+    if [ "$AIRGAP" = "1" ]; then
+        if ! addon_fetch_airgap "$addon" "$version" ; then
+            return 1
+        fi
+    else
+        if ! addon_fetch "$addon" "$version" ; then
+            return 1
+        fi
+    fi
+}
+
+# rook_upgrade_prompt_missing_images prompts the user to run the command to load the images on all
+# remote nodes before proceeding.
+function rook_upgrade_prompt_missing_images() {
+    local from_version="$1"
+    local to_version="$2"
+
+    local images_list=
+
+    if rook_upgrade_is_version_included "$from_version" "$to_version" "1.4" ; then
+        images_list="$(rook_upgrade_list_rook_ceph_images_in_manifest_file "addons/rookupgrade/10to14/Manifest")"
+    fi
+
+    if [ "$(rook_upgrade_compare_rook_versions "$to_version" "1.4")" = "1" ]; then
+        local step=
+        while read -r step; do
+            if [ -z "$step" ]; then
+                continue
+            fi
+            images_list="$(rook_upgrade_merge_images_list \
+                "$images_list" \
+                "$(rook_upgrade_list_rook_ceph_images_in_manifest_file "addons/rook/$step/Manifest")" \
+            )"
+        done <<< "$(rook_upgrade_step_versions "ROOK_STEP_VERSIONS[@]" "$(rook_upgrade_max_rook_version "1.4" "$from_version")" "$to_version")"
+    fi
+
+    if [ -z "$images_list" ]; then
+        return
+    fi
+
+    local this_hostname=
+    this_hostname=$(hostname)
+
+    local node_missing_images=
+    # shellcheck disable=SC2086
+    node_missing_images=$("$DIR"/bin/kurl cluster nodes-missing-images $images_list --exclude_host "$this_hostname")
```

Comment:
This function might not be as reliable as we'd like
it's ....okay... when searching for one image as a signpost, but k8s only reports a limited number of images per node even if there are more

Reply:
See https://github.com/replicatedhq/kURL/pull/3787/commits/9689221be54d4631eb81cb84c6ce7ab5ae77dbbd